client-core/src/wallet/syncer.rs

```diff
@@ -0,0 +1,546 @@
+                        return Err(Error::new(
+                            ErrorKind::VerifyError,
+                            "state app hash don't match block header",
+                        ));
+                    }
+                }
+                app_hash = Some(
+                    state.compute_app_hash(
+                        block_result
+                            .transaction_ids()
+                            .chain(|| (ErrorKind::VerifyError, "verify block results"))?,
+                    ),
+                );
+                if self.env.enable_fast_forward {
+                    if let Some(block) = self.fast_forward_status(&status)? {
+                        // Fast forward to latest state if possible
+                        self.handle_batch((batch, block).into())?;
+                        return Ok(());
+                    }
+                }
+
+                let block = prepare_block_header(&self.wallet, &block, &block_result)?;
+                self.update_progress(block.block_height);
+                batch.push(block);
+            }
+            if let Some(non_empty_batch) = NonEmpty::new(batch) {
+                self.handle_batch(non_empty_batch)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Fast forwards state to given status if app hashes match
+    fn fast_forward_status(&self, status: &Status) -> Result<Option<BlockHeader>> {
+        let current_app_hash = status
+            .sync_info
+            .latest_app_hash
+            .ok_or_else(|| Error::new(ErrorKind::TendermintRpcError, "latest_app_hash not found"))?
+            .to_string();
+
+        if current_app_hash == self.sync_state.last_app_hash {
+            let current_block_height = status.sync_info.latest_block_height.value();
+
+            let block = self.env.client.block(current_block_height)?;
+            let block_result = self.env.client.block_results(current_block_height)?;
+
+            Ok(Some(prepare_block_header(
+                &self.wallet,
+                &block,
+                &block_result,
+            )?))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Fast forwards state to given block if app hashes match
+    fn fast_forward_block(&mut self, block: &Block) -> Result<Option<BlockHeader>> {
+        let current_app_hash = block
+            .header
+            .app_hash
+            .err_kind(ErrorKind::TendermintRpcError, || "app_hash not found")?
+            .to_string();
+
+        if current_app_hash == self.sync_state.last_app_hash {
+            let current_block_height = block.header.height.value();
+            let block_result = self.env.client.block_results(current_block_height)?;
+            Ok(Some(prepare_block_header(
+                &self.wallet,
+                &block,
+                &block_result,
+            )?))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+fn decrypt_staking_transactions(
```

Comment:
There's no decryption in these, as a staking address is plain

Reply:
Oh, I mistakenly took the `unencrypted` as `decrypted`, I'll change it.

Reply:
Done, changed to `filter_staking_transactions`.